docs/proposals/20210310-opt-in-autoscaling-from-zero.md

```diff
@@ -0,0 +1,247 @@
+  These node groups do not have a direct relation to specific CRDs within Kubernetes, and may be handled
+  differently by each autoscaler cloud implementation. In the case of Cluster API, node groups correspond
+  directly to MachineSets and MachineDeployments that are marked for autoscaling.
+
+Refer to the [Cluster API Book Glossary](https://cluster-api.sigs.k8s.io/reference/glossary.html).
+
+## Summary
+
+The [Kubernetes cluster autoscaler](https://github.com/kubernetes/autoscaler) currently supports
+scaling on Cluster API deployed clusters. One feature that is missing from this integration is
+the ability to scale down to, and up from, a MachineSet or MachineDeployment with zero replicas.
+
+This proposal defines an opt-in mechanism whereby Cluster API users and cloud providers can define
+the specific resource requirements for each Infrastructure Machine Temalate they create. These
+requirements are then utilized by the cluster autoscaler to make decisions about the number
+of nodes to add during scale up operations.
+
+## Motivation
+
+Allowing the cluster autoscaler to scale down its node groups to zero replicas is a common feature
+implemented for many of the integrated cloud providers. It is a popular feature that has been
+requested for Cluster API on multiple occasions. This feature empowers users to reduce their
+operational resource needs, and likewise reduce their operating costs.
+
+Given that Cluster API is an abstraction point that provides access to multiple concrete cloud
+implementations, this feature might not make sense in all scenarios. To accomodate the wide
+range of deployment options in Cluster API, the scale to zero feature will be optional for
+users and cloud providers.
+
+### Goals
+
+- Provide capability for Cluster API MachineSets and MachineDeployments to scale from and to zero replicas.
+- Provide a mechanism for users to override the defaults for any given MachineSet or MachineDeployment.
+- Create an optional API contract that defines the minimum needed information for scaling from and to zero replicas.
+
+### Non-Goals/Future Work
+
+- Create an API contract that cloud providers must follow.
+- Create an API that replicates Taint and Label information from Machines to MachineSets and MachineDeployments.
+- Create additional CRDs or controllers
+
+## Proposal
+
+To facilitate scaling from zero replicas, the minimal information needed by the cluster autoscaler
+is the CPU and memory resources for the target node group that will be scaled. The autoscaler uses
+this information to create a prediction about how many nodes should be created when scaling. In
+most sitatuations this information can be directly read from the nodes that are running within a
+node group. But, during a scale from zero sitatuation (ie when a node group has zero replicas) the
+autoscaler needs to acquire this information from the cloud provider.
```

Comment:
the problem is more on the autoscaler side with how it does calculations about scaling up. doing the _1 to n_ scaling is easier because the autoscaler can evaluate the resources used by the nodes running to calculate how many nodes should be added.
when the autoscaler sees a node group with size 0, it needs some information to instruct about how to calculate the number of needed nodes to satisfy the unsheduled pods. for the other providers, these are just lookup tables based on the instance types in the node group. but for capi the linkage is not as straightforward.

Reply:
> it needs some information to instruct about how to calculate the number of needed nodes to satisfy the unsheduled pods

I don't have direct knowledge of the autoscaler internals, but what is kind of strange, is that because the autoscaler can't manage "lack of informations", we are spreading autoscaler concerns across other APIs.

If the autoscaler could "fallback to 1" when there are no informations there will be a much more cleaner separation of concerns...

Reply:
> I don't have direct knowledge of the autoscaler internals, but what is kind of strange, is that because the autoscaler can't manage "lack of informations", we are spreading autoscaler concerns across other APIs.

i think this is impacting us more directly than any other cloud provider for the autoscaler though. mainly due to our architecture. for example, the other cloud providers have implementations that live within the autoscaler repository. so for them, this information is compiled directly into the autosacler (eg AWS has a lookup table with this information).

> If the autoscaler could "fallback to 1" when there are no informations there will be a much more cleaner separation of concerns...

it might be possible for us to cache this information inside our implementation at the autoscaler, but i would worry about it getting lost on a restart. i'll give this some thought.

Reply:
>If the autoscaler could "fallback to 1" when there are no informations there will be a much more cleaner separation of concerns...

Autoscaler falling back to 1 would completely defeat the purpose of the feature since you only want to scale up if the upcoming nodes are a good fit for the workloads to be scheduled.
The contract for any provider to be autoscalable from zero is to provide this information therefore the need to expose it so it can be consumed.

Reply:
Fall back to 1 will serve only for moving the autoscaler out of a situation that it can manage properly (it is required to scale up, but there is not enough info to determine how many machine do we actually need).

My assumption, is that immediately after scaling to 1, we are back in a situation the autoscaler is already capable to manage properly, and thus the existing logic will kick in and add further nodes as required.
But might be am I wrong..
@elmiko to confirm/deny

Reply:
> My assumption, is that immediately after scaling to 1, we are back in a situation the autoscaler is already capable to manage properly, and thus the existing logic will kick in and add further nodes as required.

this is correct. the autoscaler core essentially uses an interface to ask the provider-specific bits for a "node group template" (a template of what the instances in that node group look like). when a node group is at zero we will use the information from the resource structure, but if the node group has a size greater than zero, then the autoscaler will use a node group template based on the nodes that already exist.

Reply:
thanks for context.
the idea is to have something similar:

- if the replicas is 0, and it is required to scale up:
   - start scaling to 1 (given that we don't know how many machines are actually required due to lack of machines capacity info) 
   
then, as soon as the machine capacity are available, the autoscaler will determine if to add more replicas (as of today)

Reply:
that is more or less how it works